src/dstack/_internal/core/services/ssh/attach.py

```diff
@@ -1,14 +1,38 @@
+import atexit
+import re
+import subprocess
 import time
+from typing import Optional, Tuple
 
 from dstack._internal.core.errors import SSHError
 from dstack._internal.core.services.configs import ConfigManager
 from dstack._internal.core.services.ssh.ports import PortsLock
 from dstack._internal.core.services.ssh.tunnel import ClientTunnel
 from dstack._internal.utils.path import PathLike
-from dstack._internal.utils.ssh import include_ssh_config, update_ssh_config
+from dstack._internal.utils.ssh import get_ssh_config, include_ssh_config, update_ssh_config
 
 
 class SSHAttach:
+    @staticmethod
+    def reuse_control_sock_path_and_port_locks(run_name: str) -> Optional[Tuple[str, PortsLock]]:
```

Comment:
I see many edge cases in this approach.

What happens if we reuse the connection and exit early? I suppose we will close the tunnel with atexit and the original process will fail.

We could reuse the control path socket (for example, under `~/.dstack/sock/<job-name>`), but forward only a single port to the runner logs to a free local port.